### PR TITLE
Handle thread crash/abort even when main thread is blocking (busy-waiting). NFC

### PR DESCRIPTION
**Pull request by sbc100**
Previously if a thread crashed when the main thread was blocked on
something (such as pthread_join) it would never get the `onerror`
message because it would never return the event loop.

Without this change the main thread will simple loop forever after the
secondary thread crashes.

### CODE REVIEW

**Comment by sbc100** [comment]
Still working a on test ..

File: system/lib/pthread/emscripten_futex_wait.c [hunk 1]
```diff
@@ -106,11 +108,24 @@ static int futex_wait_busy(volatile void *addr, uint32_t val, double timeout) {
...
   if ((((intptr_t)addr)&3) != 0) {
     return -EINVAL;
   }
 
+  // When a secondary thread crashes we need to be able to interrupt the main
+  // thread is its in a busy loop.  We want to avoid using the normal proxying
+  // mechanism to send this message since it can allocate (or otherwise itself
+  // crash) so use a low level atomic primitive for this signal.
+  if (emscripten_is_main_runtime_thread() && thread_crashed)
+    abort();
```
[the remark this answers is not in the document]
Comment (sbc100) [reply system/lib/pthread/emscripten_futex_wait.c]: Yes, this is like one of the threads segfaulting.

If the main thread is returning to the even loop there is no issue.  This patch is only for the case when the main thread is blocking/looping and not returning (e.g. pthread_join).

File: system/lib/pthread/emscripten_futex_wait.c [hunk 2]
```diff
@@ -106,11 +109,24 @@ static int futex_wait_busy(volatile void *addr, uint32_t val, double timeout) {
...
+  thread_crashed = true;
+}
+
 int emscripten_futex_wait(volatile void *addr, uint32_t val, double max_wait_ms) {
   if ((((intptr_t)addr)&3) != 0) {
     return -EINVAL;
   }
 
+  // When a secondary thread crashes we need to be able to interrupt the main
+  // thread is its in a busy loop.  We want to avoid using the normal proxying
```
Comment (tlively) [inline system/lib/pthread/emscripten_futex_wait.c:124]:
```suggestion
  // thread if it's in a busy loop.  We want to avoid using the normal proxying
```

**Review by tlively — APPROVED** [review]
Aha, that makes sense now. Checking for crashes in futex_wait is sufficient because we're assuming that the main thread will call pthread_join at some point.

**Comment by sbc100** [comment]
> Aha, that makes sense now. Checking for crashes in futex_wait is sufficient because we're assuming that the main thread will call pthread_join at some point.

All our block primitives bottom out here in futex_wait.  If the thread is busy in a compute loop we can't do anything.. but such a program has worse problems to deal with.

**Comment by sbc100** [comment]
See https://github.com/emscripten-core/posixtestsuite/issues/10

File: system/lib/pthread/emscripten_futex_wait.c [hunk 3]
```diff
@@ -106,11 +109,28 @@ static int futex_wait_busy(volatile void *addr, uint32_t val, double timeout) {
...
+  // When a secondary thread crashes, we need to be able to interrupt the main
+  // thread even if it's in a blocking/looping on a mutex.  We want to avoid
+  // using the normal proxying mechanism to send this message since it can
+  // allocate (or otherwise itself crash) so use a low level atomic primitive
+  // for this signal.
+  if (emscripten_is_main_runtime_thread() && thread_crashed) {
+    // Return the event loop so we can handle the message from the crashed
+    // thread.
+    emscripten_unwind_to_js_event_loop();
+  }
```
[the remark this answers is not in the document]
Comment (kripken) [reply system/lib/pthread/emscripten_futex_wait.c:132]: Oh, great! Very nice.

**Pull request merged**